File: quad_state_estimation/test/quaternion_helper.py
#!/usr/bin/env python

#
# AER 1810 Quadrotor Landing Project
# Quaternion Helper Functions
#

from __future__ import division, print_function, absolute_import

import numpy as np

small_angle_tol = 1E-10

def quaternion_exp(pure_quat):
    "Quaternion exponential map from pure to unit quaternion"
    unit_quat = np.zeros(4)
    norm = np.linalg.norm(pure_quat)

    # Real part is fine
    unit_quat[3] = np.math.cos(norm / 2)

    if norm < small_angle_tol:
        # Small rotation, use Taylor series approximation for vector part of quaternion to avoid zero division
        unit_quat[0:3] = pure_quat.flatten() / 2 * (1 - (norm ** 2) / 24)
    else:
        # Rotation is large enough, use true rotation axis
        unit_quat[0:3] = pure_quat.flatten() / norm * np.math.sin(norm / 2)
        
    return unit_quat

def quaternion_log(unit_quat):
    "Quaternion logarithmic map from unit to pure quaternion"
    pure_quat = np.zeros(3)

    norm = np.linalg.norm(unit_quat[0:3])

    if norm < small_angle_tol:
        # Approximate map with Taylor series for atan
        pure_quat = 2 / unit_quat[3] * (1 - (norm / unit_quat[3]) ** 2 / 3)*unit_quat[0:3]
    else:
        # Large angles, use full version of atan
        phi = 2 * np.math.atan2(norm, unit_quat[3])
        u = unit_quat[0:3] / norm
        pure_quat = phi * u
    
    return pure_quat

def quaternion_norm(unit_quat):
    "Normalize unit length quaternion and clip to single cover range"
    q_w_lim = -0.75

    unit_quat_norm = unit_quat / np.linalg.norm(unit_quat)

    if unit_quat_norm[3] < q_w_lim:
        return -unit_quat_norm
    else:
        return unit_quat_norm

def identity_quaternion():
    "Identity quaternion"
    q = np.zeros(4)
    q[3] = 1
    return q

def skew_symm(vector):
    "Return the 3x3 skew symmetric matrix of a vector"
    vector_flatten = vector.flatten()
    skew = np.array([[0,-vector_flatten[2], vector_flatten[1]],
                    [vector_flatten[2], 0, -vector_flatten[0]],
                    [-vector_flatten[1], vector_flatten[0], 0]])
    return skew

def vex_symm(skew):
    "Return the vex operator of a skew symmetric matrix"
    vex = np.zeros((3,1))
    vex[0,0] = (skew[2,1] - skew[1,2]) / 2
    vex[1,0] = (skew[0,2] - skew[2,0]) / 2
    vex[2,0] = (skew[1,0] - skew[0,1]) / 2
    return vex

def exponential_map(vector):
    "Return the rotation matrix corresponding to a Lie algebra vector"
    norm = np.linalg.norm(vector)

    if norm < small_angle_tol:
        # First order approximation
        return np.eye(3) + skew_symm(vector)
    else:
        # Full exponential map
        return (np.math.cos(norm) * np.eye(3) +
        (1 - np.math.cos(norm)) * np.dot(vector.reshape((3,1)), vector.reshape((1,3))) +
        np.math.sin(norm) * skew_symm(vector))

File: quad_state_estimation/src/quaternion_helper.cpp
/*
    AER 1810 Quadrotor Landing Project
    Quaternion Helper Functions
*/

#include <quaternion_helper.h>

constexpr double kSmallRotationTolerance{1E-10};

// Quaternion exponential map from pure to unit quaternion
Eigen::Quaterniond quaternion_exp(Eigen::VectorXd pure_quat) {
    Eigen::Quaterniond unit_quat;

    double norm = pure_quat.norm();

    // Real part is fine
    unit_quat.w() = cos(norm/2);

    if (norm < kSmallRotationTolerance)
    {
        // Small rotation, use Taylor series approximation for vector part
        unit_quat.vec() = pure_quat/2*(1-pow(norm,2)/24);
    }
    else
    {
        // Rotation is large enough, use true rotation axis
        unit_quat.vec() = pure_quat/norm*sin(norm/2);
    }

    quaternion_norm(unit_quat);

    return unit_quat;
}

// Quaternion logarithmic map from unit to pure quaternion
Eigen::VectorXd quaternion_log(Eigen::Quaterniond unit_quat) {
    Eigen::VectorXd pure_quat = Eigen::VectorXd::Zero(3);
    Eigen::VectorXd quat_vec = Eigen::VectorXd::Zero(3);

    quat_vec << unit_quat.x(),unit_quat.y(),unit_quat.z();
    double vec_norm = quat_vec.norm();

    if (vec_norm < kSmallRotationTolerance)
    {
        // Approximate map with Taylor series for atan
        pure_quat = 2/unit_quat.w()*(1-pow(vec_norm/unit_quat.w(),2)/3)*quat_vec;
    }
    else
    {
        // Large angles, use full version of atan
        double phi = 2*atan2(vec_norm,unit_quat.w());
        pure_quat = phi/vec_norm*quat_vec;
    }
    
    return pure_quat;
}

// Normalize unit length quaternion and clip to single cover range
void quaternion_norm(Eigen::Quaterniond &unit_quat) {
    double q_w_lim = -0.75;
    unit_quat.normalize();

    if (unit_quat.w()<q_w_lim)
    {
        unit_quat.w() = -unit_quat.w();
        unit_quat.x() = -unit_quat.x();
        unit_quat.y() = -unit_quat.y();
        unit_quat.z() = -unit_quat.z();
    }
}

// Return the 3x3 skew symmetric matrix of a vector
Eigen::MatrixXd skew_symm(Eigen::VectorXd vector) {
    Eigen::MatrixXd mat(3,3);
    
    mat << 0, -vector(2), vector(1),
        vector(2), 0 , -vector(0),
        -vector(1), vector(0), 0;
    
    return mat;
}

// Return the 4x1 vector corresponding to a quaternion
Eigen::VectorXd quat_to_vec(Eigen::Quaterniond quaternion) {
    Eigen::VectorXd vec = Eigen::VectorXd::Zero(4);
    vec << quaternion.x(),quaternion.y(),quaternion.z(),quaternion.w();
    return vec;
}

// Return the quaternion corresponding to a 4x1 vector
Eigen::Quaterniond vec_to_quat(Eigen::VectorXd vec) {
    Eigen::Quaterniond quat(vec(3),vec(0),vec(1),vec(2));
    return quat;
}

// Return the vex operator of a skew symmetric matrix
Eigen::Vector3d VexSymm(const Eigen::Matrix3d& skew_symm) {
    return Eigen::Vector3d{(skew_symm[2,1] - skew_symm[1,2]) / 2,
                           (skew_symm[0,2] - skew_symm[2,0]) / 2,
                           (skew_symm[1,0] - skew_symm[0,1]) / 2};
}

// Return the rotation matrix corresponding to a Lie algebra vector
Eigen::Matrix3d exponential_map(const Eigen::Vector3d& lie_vector) {
    const double norm = lie_vector.norm();

    // First order approximation for small angles
    if (norm < kSmallRotationTolerance) return (Eigen::Matrix3d::Identity()).colwise() + skew_symm(lie_vector);

    // Full exponential map
    const Eigen::Vector3d vector_norm = lie_vector / norm;

    return (cos(norm) * Eigen::Matrix3d::Identity() +
    (1 - cos(norm)) * vector_norm * vector_norm.transpose() + 
    sin(norm) * skew_symm(vector_norm));
}